doc(snap): packing your ros2 package as snap (#2358)
ros2/ros2_documentation: source/Tutorials/Packing-your-ROS2-package-as-a-snap.rst

> .. _CreateSnap:

Reviewer: Since we are now preferring to use the `:doc:` tag to refer to pages, we don't generally need these anchors anymore.  Just remove this.

> Packing your ROS2 package as a snap

Reviewer: ```suggestion
Packaging your ROS 2 package as a snap
```

Reviewer: We prefer "ROS 2" over "ROS2", so please make that fix throughout this document.

===========================================

.. contents:: Table of Contents
   :depth: 2
   :local:

What are snaps?
---------------

> `Snaps <https://snapcraft.io/>`_ are containers that bundle an application and all its dependencies.They

Reviewer: Please make one sentence per line, please.  That applies to the rest of the document as well.

offer several features that address important concerns as one gets closer to
shipping a robotic platform:

- Container solution: Snaps bundle all your dependencies and assets in one
  package (including ROS2) making your application installable on dozens of
  Linux distributions and across distro versions.
- Strict confinement: Snaps are designed to be `secure and isolated <https://snapcraft.io/docs/snap-confinement>`_ from the
  underlying system and other applications, with `dedicated interfaces <https://snapcraft.io/docs/supported-interfaces>`_ to
  access the host machine.
- Managing updates: Snaps can update `automatically and transactionally <https://snapcraft.io/docs/keeping-snaps-up-to-date>`_, making
  sure the device is never broken.
- Release management: Snaps' `multiple release channels <https://snapcraft.io/docs/channels>`_ allow you to have
  role-based access controls and application versioning,
  making A/B testing easy and releasing fixes faster.

Creating a snap
---------------

> This tutorial will demonstrate how to use `Snapcraft <https://github.com/snapcore/snapcraft>`_ to create a new snap, and then the usage.

Reviewer: ```suggestion
This tutorial will demonstrate how to use `Snapcraft <https://github.com/snapcore/snapcraft>`_ to create a new snap, and then how to use it.
```

First, install Snapcraft. It's recommended to install it from the Snap Store:

.. code-block:: bash

    sudo snap install --classic snapcraft

(Note that using the apt repositories for snapcraft is not recommended and
this tutorial will assume that you installed the snap.)

Snapcraft has built-in support for ``Colcon``.
You point it to your package, and tell it what commands to include in the snap.

> For our example, we will use demo_nodes_cpp from the `ros2_demos <https://github.com/ros2/demos/tree/foxy>`_.

Reviewer: To make this work across ROS 2 distributions, we can use some macros here:

```suggestion
For our example, we will use demo_nodes_cpp from the `ROS 2 demos <https://github.com/ros2/demos/tree/{REPOS_FILE_BRANCH}>`_.
```

Initialize a new Snapcraft project here:

.. code-block:: bash

>     Snapcraft init

Reviewer: Is this really a capital `S`?  If so, that's fine, it's just unusual for a Unix command-line program.

This will create a file in a subdirectory ``snap/snapcraft.yaml``.

The snapcraft file
^^^^^^^^^^^^^^^^^^

Open that ``snap/snapcraft.yaml`` file and make it look like this:

.. code-block:: yaml

    name: ros2-talker-listener
    version: '0.1'
    summary: ROS2 Talker/Listener example
    description: |
      This example launches a ROS2 talker and listener.

    confinement: devmode
    base: core20

    parts:
      ros-demos:
        plugin: colcon
        source: https://github.com/ros2/demos.git

>         source-branch: foxy

Reviewer: ```suggestion
        source-branch: {ROS_DISTRO}
```

        colcon-packages: [demo_nodes_cpp]
        build-packages: [make, gcc, g++]
        stage-packages: [ros-foxy-ros2launch]

    apps:
      ros2-talker-listener:

>       command: opt/ros/foxy/bin/ros2 launch demo_nodes_cpp talker_listener.launch.py

Reviewer: Should this be:

`command: /opt/ros/foxy/bin/ros2 launch demo_nodes_cpp talker_listener.launch.py`?  I'm not familiar with snaps so I don't know.

Reviewer 2: Nop, this is a relative path resolved by the snap.
